polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1716,6 +1717,268 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
+      return failure();
+    auto range = *setNDRangeOps.begin();
+
+    // Extract the number of parameters and the pointer to the
+    // `kernel_signatures` data structure from annotations in the current block.
+    auto failureOrParamInfo = getKernelParameterInfo(op->getBlock());
+    if (failed(failureOrParamInfo))
+      return failure();
+    auto [numParams, paramDesc] = *failureOrParamInfo;
+
+    // Parse `kernel_signatures` to get the parameter kinds.
+    auto failureOfParamKinds = getParameterKinds(numParams, paramDesc);
+    if (failed(failureOfParamKinds))
+      return failure();
+    auto paramKinds = *failureOfParamKinds;
+
+    // Adjust `num_params` if the last entry is a sentinel.
+    if (numParams > 0 && paramKinds.back() == kernel_param_kind_t::kind_invalid)
+      --numParams;
+
+    // No point in trying to raise a `schedule_kernel` without arguments.
+    if (numParams == 0)
+      return failure();
+
+    // Discover `set_captured` ops in the CGF. We check that we have captured a
+    // value matching the expected kind for each parameter.
+    auto setCapturedOps = enclosingFunc.getOps<sycl::SYCLHostSetCaptured>();
+    unsigned numCaptured = llvm::range_size(setCapturedOps);
+
+    // Wrong number of `set_captured` ops.
+    if (numCaptured != numParams)
+      return failure();
+
+    SmallVector<Value> args(numCaptured);
+    SmallVector<Type> syclTypes(numCaptured, rewriter.getNoneType());
+    for (auto capture : setCapturedOps) {
+      unsigned idx = capture.getIndex();
+      // Index is out of bounds.
+      if (idx >= numCaptured)
+        return failure();
+
+      args[idx] = capture.getValue();
+
+      auto syclType = capture.getSyclType();
+      auto kind = paramKinds[idx];
+      switch (kind) {
+      case kernel_param_kind_t::kind_accessor:
+        // We expected an accessor, but the captured value isn't one.
+        if (!syclType.has_value() || !isa<sycl::AccessorType>(syclType.value()))
+          return failure();
+
+        syclTypes[idx] = syclType.value();
+        break;
+
+      case kernel_param_kind_t::kind_std_layout:
+        // We expected standard layout, but the captured value is a
+        // SYCL-specific type.
+        if (syclType.has_value())
+          return failure();
+        break;
+
+      default:
+        // TODO: support other argument kinds as well
+        return failure();
+      }
+    }
+
+    // `set_captured` ops didn't cover all args.
+    if (!args.empty() &&
+        !llvm::all_of(args, [](auto v) { return static_cast<bool>(v); }))
+      return failure();
+
+    // Finally construct the op with the raised information.
+    rewriter.replaceOpWithNewOp<sycl::SYCLHostScheduleKernel>(
+        op, handler, op.getKernelName(), range.getRange(), range.getOffset(),
+        args, rewriter.getTypeArrayAttr(syclTypes), range.getNdRange());
+
+    return success();
+  }
+
+private:
+  // Copy of `sycl::_V1::detail::kernel_param_kind_t`.
```

Comment:
Can we mention the header where this is coming from?

Reply:
Done.